README
Shoutstats

Introduction
~~~~~~~~~~~~
Shoutstats is a fast, free Shoutcast server statistic analysis program.
It produces instant and dynamic usage reports in HTML format, for viewing in a
standard browser.

Shoutstats is a bunch of php scripts and a rrdtools database.

It has been written under a Debian GNU/Linux.

That's all for now !

Required
~~~~~~~~
- PHP 4
- rrdtools
- One or more shoutcast servers
- a crontab
- wget

License
~~~~~~~
Shoutstat - let's humanize your shoutcast stats
Copyright (C) 2000,2004 Laurent Raufaste

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA

Contacts and website
~~~~~~~~~~~~~~~~~~~~
The main webpage for shoutstats is http://www.glop.org/shoutstats
Please send any suggestions, bugfixes, or anything else to analogue@glop.org


CHANGES
shoutstats 0.8.5
- Add support to icecast2 servers

shoutstats 0.8.4
- Fixed a bug occuring under a default PHP5 install

shoutstats 0.8.3
- XHTML/1.1 and CSS2 compliant
- Added a new configuration file to set up your station name and the rrdtool
  path
- Some cosmetic changes

shoutstats 0.8.2:
- updated some links
- rrdtool binary can now be defined

shoutstats 0.8.1:
- fixed a stupid bug which prevented shoutstats to get any the stat at all.

shoutstats 0.8.0:
- fixed a bug link bug under Opera.
- fixed the bug which would reset the total graph after adding a new server
- changed the format of the configuration file
- added the possibility to name the servers
- cosmetic changes

shoutstats 0.7.1:
- fixed the listener count when a server is down

shoutstats 0.7:
- rrdtools database instead of SQL
- rewritten everything around rrdtools

shoutstats 0.6:
- internal release

shoutstats 0.5:
- password no longer needed to get stats from the servers
- code cleaning
- public release
- added month and year stats
- added a setup interface

shoutstats 0.4:
- SQL engine
- Internal release

shoutstats 0.3:
- Enhanced portability
- Bugfixes

shoutstats 0.2:
- Added support for 99+ simultaneous listeners

shoutstats 0.1:
- Initial release, works fine for me.


sslib.php
<?php
// Shoutstats version
define('SS_VERSION',	'0.8.5');

// The file providing the the config variables
define('SS_CONF',	'config.ini');

// The file providing the server list
define('SS_SERVERS',	'servers.ini');

// The folder containing the generated PNG
define('SS_PATH_GFX',	'gfx');

// The folder containing the RRDtool databases
define('SS_PATH_RRD',	'rrd');

if (!file_exists(SS_CONF))
{
	die(SS_CONF.' file is missing');
}
$configuration = parse_ini_file(SS_CONF, TRUE);
define('SS_RRDTOOL_COMMAND', $configuration['Configuration']['rrdtool_command']);
define('SS_NAME', $configuration['Configuration']['stream_name']);

//if(!is_executable(SS_RRDTOOL_COMMAND))
//	die(SS_RRDTOOL_COMMAND . ' not found. Please update the ' . SS_CONF . ' file');
exec(SS_RRDTOOL_COMMAND, $output);
if(!count($output))
{
	die(SS_RRDTOOL_COMMAND . ' not found. Please update the ' . SS_CONF . ' file');
}

//
// Get the server list from the configuration file
//

function GetServerList()
{
	if (!file_exists(SS_SERVERS))
	{
		die(SS_SERVERS.' file is missing');
	}

	$servers = parse_ini_file(SS_SERVERS, TRUE);
	ksort($servers);

	if (!count($servers))
	{
		die('no server configured, edit the '.SS_CONF.' file');
	}

	return $servers;
}

//
// Get the number of current and max listeners from the specified server
//

function GetShoutcastStats($host, $port)
{
  $fp = fsockopen($host, $port, $errno, $errstr, 30);

  // can't connect =(
  if (!$fp)
  {
	print("$errstr ($errno)<br>\n");
	$server['current'] = 0;
	$server['max'] = 0;
  // oh yes, it can connect
  }
  else
  {
      fputs($fp, "GET /7.html HTTP/1.0\r\nUser-Agent: Mozilla/4.0 (compatible; MSIE 5.5; Windows NT 5.0)\r\n\r\n");
      while (!feof($fp))
	  {
          $content .= fgets($fp,128);
      }
      fclose($fp);

      $debut = strpos($content, '<body>') + strlen('<body>');
      $fin = strpos($content, '</body>', $debut);
      $string = substr($content, $debut, $fin - $debut);

	  $stats = explode(',', $string);

	// server is up but no source is connected
	if ($stats[1] == 0)
	{
		$server['current'] = 0;
	// everything is ok
	}
	else
	{
		$server['current'] = $stats[0];
	}

	$server['max'] = $stats[3];
  }

	// debug
	// print("$host:$port = {$server['current']}/{$server['max']}\n");
	return $server;
}

function GetIcecastStats($host,$port,$mp)
{
  $fp = fsockopen($host, $port, $errno, $errstr, 30);

  // can't connect =(
  if (!$fp) {
        print("$errstr ($errno)<br>\n");
        $server['current'] = 0;
        $server['max'] = 0;
  // oh yes, it can connect
  } else {
      fputs($fp, "GET /status2.xsl?mount=/".$mp." HTTP/1.0\r\nUser-Agent: Mozilla/4.0 (compatible; MSIE 5.5; Windows NT 5.0)\r\n\r\n");
      while (!feof($fp)) {
          $content .= fgets($fp,128);
      }
      fclose($fp);

      $debut = strpos($content, $mp) + strlen($mp);
      $fin = strpos($content, '</pre>', $debut);
      $string = substr($content, $debut, $fin - $debut);
      $stats = explode(',', $string);
      $server['current'] = $stats[3];

  }
        // debug
        //print("{$string}\n");
        //print("$host:$port = {$server['current']}\n");
  return $server;
}


//
// Template used to display each generated graph
//

function DisplayGraph($txt_freq, $rrdgfx)
{
	$size = GetImageSize($rrdgfx);
?>
<p>
<?=$txt_freq?> graph:<br />
<img src="<?=$rrdgfx?>" <?=$size[3]?> alt="<?=$txt_freq?> graph" /><br />
</p>
<?
}


update.php
<?php
require('sslib.php');

$servers = GetServerList();

if(!file_exists(SS_PATH_RRD))
{
	mkdir(SS_PATH_RRD,0777);
}

foreach ($servers as $n => $s)
{
	if($s["type"]=="shoutcast")
        {
                $audience[$i]=GetShoutcastStats($s["host"],$s["port"]);
                $rddfile = SS_PATH_RRD."/{$s["host"]}.".$s["port"].".rrd";
        }
        else
        {
                $audience[$i]=GetIcecastStats($s["host"],$s["port"],$s["mpoint"]);
                $rddfile = SS_PATH_RRD."/{$s["host"]}.".$s["port"].".{$s["mpoint"]}.rrd";
        }

	if(!file_exists($rddfile))
	{
		system(SS_RRDTOOL_COMMAND.
			   ' create '.$rddfile.
			   'DS:audience:GAUGE:600:U:U '.
			   'RRA:AVERAGE:0.5:1:288 '.
			   'RRA:AVERAGE:0.5:6:336 '.
			   'RRA:AVERAGE:0.5:24:420 '.
			   'RRA:AVERAGE:0.5:288:365');
	}
	system(SS_RRDTOOL_COMMAND.
		   ' update '.$rddfile.' '.
		   '-t audience '.
		   'N:'.$audience[$i]['current']);
}


index.php
<?
echo '<?xml version="1.0" encoding="UTF-8"?>' . "\n";
require('sslib.php');

$server = $_GET['server'];
$servers = GetServerList();

if(!file_exists(SS_PATH_GFX))
	mkdir(SS_PATH_GFX,0777);
?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" >
<head>
 <? if (strlen($server)) : ?>
 <title>Shoutstats for <?=SS_NAME?> - <?=$server?></title>
 <? else: ?>
 <title>Shoutstats for <?=SS_NAME?></title>
 <? endif; ?>
 <meta http-equiv="refresh" content="300" />
 <link rel="stylesheet" href="shoutstats.css" type="text/css" media="screen" />
</head>
<body>

<h1>Shoutstats for <?=SS_NAME?></h1>

<form action="<?=basename($SCRIPT_FILENAME)?>" method="get">
<fieldset>
<legend>Shoutstats Server Selection</legend>
<label for="server">Show statistics for</label>
<select name="server" id="server">
<option value="">All servers together</option>
<?
// combobox items generation of a list of server
foreach ($servers as $c_name => $c_server)
{
	if ($c_name == $server) : ?>
                <option value="<?=$c_name?>" selected><?=$c_name?> (<?=$servers[$c_name]['type']?>)</option>
        <? else: ?>
                <option value="<?=$c_name?>"><?=$c_name?> (<?=$servers[$c_name]['type']?>)</option>
        <? endif;

}
?>
</select>
<input type="submit" value="Select" />
</fieldset>
</form>

<?
// a specific server has been selected
if(strlen($server)):
if($servers[$server]['type']=='icecast') $mp=".".$servers[$server]['mpoint'];

// generate and display the 24 hours server specific graph
$txt_freq = 'hourly';
$rrdgfx=SS_PATH_GFX."/hourly-{$servers[$server]['host']}.{$servers[$server]['port']}{$mp}.png";
$rrdfile=SS_PATH_RRD."/{$servers[$server]['host']}.{$servers[$server]['port']}{$mp}.rrd";
exec(SS_RRDTOOL_COMMAND . " graph $rrdgfx --lazy -l 0 -a PNG DEF:myaudience=$rrdfile:audience:AVERAGE LINE2:myaudience#FF0000");
DisplayGraph($txt_freq, $rrdgfx);

// generate and display the 7 days server specific graph
$txt_freq = 'daily';
$rrdgfx=SS_PATH_GFX."/daily-{$servers[$server]['host']}.{$servers[$server]['port']}{$mp}.png";
exec(SS_RRDTOOL_COMMAND . " graph $rrdgfx --lazy -l 0 -a PNG -s e-1w DEF:myaudience=$rrdfile:audience:AVERAGE LINE2:myaudience#FF0000");
DisplayGraph($txt_freq, $rrdgfx);

// generate and display the 5 weeks server specific graph
$txt_freq = 'weekly';
$rrdgfx=SS_PATH_GFX."/weekly-{$servers[$server]['host']}.{$servers[$server]['port']}{$mp}.png";
exec(SS_RRDTOOL_COMMAND . " graph $rrdgfx --lazy -l 0 -a PNG -s e-5w DEF:myaudience=$rrdfile:audience:AVERAGE LINE2:myaudience#FF0000");
DisplayGraph($txt_freq, $rrdgfx);

// generate and display the 12 months server specific graph
$txt_freq = 'monthly';
$rrdgfx=SS_PATH_GFX."/monthly-{$servers[$server]['host']}.{$servers[$server]['port']}{$mp}.png";
exec(SS_RRDTOOL_COMMAND . " graph $rrdgfx --lazy -l 0 -a PNG -s e-1y DEF:myaudience=$rrdfile:audience:AVERAGE LINE2:myaudience#FF0000");
DisplayGraph($txt_freq, $rrdgfx);

// no server has been selected
else:

$rrdexecend = ' ';
$i = 0;
foreach ($servers as $n => $s)
{
	$mp=""; if($s['type']=='icecast') $mp=".".$s['mpoint'];
        $rrdfile = SS_PATH_RRD."/{$s['host']}.{$s['port']}{$mp}.rrd";
	$rrdexecend .= "DEF:myaudience$i=$rrdfile:audience:AVERAGE ";
	$rrdexecend .= "CDEF:audience$i=myaudience$i,UN,0,myaudience$i,IF ";
	$i++;
}
if(count($servers)>1) {
	$rrdexecend .= 'CDEF:mytotal=';
	for($i=0; $i<count($servers); $i++)
		$rrdexecend .= "audience$i,";
	for($i=0; $i<count($servers)-2; $i++)
		$rrdexecend .= '+,';
	$rrdexecend .= '+ ';
} else {
	$rrdexecend .= 'CDEF:mytotal=audience0 ';
}
$rrdexecend .= 'LINE2:mytotal#FF0000';

// generate and display the 24 hours graph
$txt_freq = 'hourly';
$rrdgfx = SS_PATH_GFX.'/hourly-all.png';
$rrdexecstart = SS_RRDTOOL_COMMAND . " graph $rrdgfx --lazy -l 0 -a PNG";
exec($rrdexecstart.$rrdexecend);
DisplayGraph($txt_freq, $rrdgfx);

// generate and display the 7 days graph
$txt_freq = 'daily';
$rrdgfx = SS_PATH_GFX.'/daily-all.png';
$rrdexecstart = SS_RRDTOOL_COMMAND . " graph $rrdgfx --lazy -l 0 -a PNG -s e-1w";
exec($rrdexecstart.$rrdexecend);
DisplayGraph($txt_freq, $rrdgfx);

// generate and display the 5 weeks graph
$txt_freq = 'weekly';
$rrdgfx = SS_PATH_GFX.'/weekly-all.png';
$rrdexecstart = SS_RRDTOOL_COMMAND . " graph $rrdgfx --lazy -l 0 -a PNG -s e-5w";
exec($rrdexecstart.$rrdexecend);
DisplayGraph($txt_freq, $rrdgfx);

// generate and display the 12 months graph
$txt_freq = 'monthly';
$rrdgfx = SS_PATH_GFX.'/monthly-all.png';
$rrdexecstart = SS_RRDTOOL_COMMAND . " graph $rrdgfx --lazy -l 0 -a PNG -s e-1y";
exec($rrdexecstart.$rrdexecend);
DisplayGraph($txt_freq, $rrdgfx);

endif;
?>

<p>
<a href="http://www.gnu.org/copyleft/gpl.html">GPL</a>
<a href="http://www.glop.org/shoutstats/">Shoutstats</a> 
<a href="CHANGES"><?=SS_VERSION?></a> -
<a href="http://validator.w3.org/check?uri=referer">XHTML/1.1</a> and 
<a href="http://jigsaw.w3.org/css-validator/check/referer">CSS2</a> compliant
</p>

</body>
</html>
